File: itsm-backend/ent/schema/change.go
package schema

import (
	"time"

	"entgo.io/ent"
	"entgo.io/ent/schema/field"
)

// Change holds the schema definition for the Change entity.
type Change struct {
	ent.Schema
}

// Fields of the Change.
func (Change) Fields() []ent.Field {
	return []ent.Field{
		field.String("title").
			Comment("变更标题").
			NotEmpty(),
		field.Text("description").
			Comment("变更描述").
			Optional(),
		field.Text("justification").
			Comment("变更理由").
			Optional(),
		field.String("type").
			Comment("变更类型").
			Default("normal"),
		field.String("status").
			Comment("状态").
			Default("draft"),
		field.String("priority").
			Comment("优先级").
			Default("medium"),
		field.String("impact_scope").
			Comment("影响范围").
			Default("medium"),
		field.String("risk_level").
			Comment("风险等级").
			Default("medium"),
		field.Int("assignee_id").
			Comment("处理人ID").
			Optional(),
		field.Int("created_by").
			Comment("创建人ID").
			Positive(),
		field.Int("tenant_id").
			Comment("租户ID").
			Positive(),
		field.Time("planned_start_date").
			Comment("计划开始时间").
			Optional(),
		field.Time("planned_end_date").
			Comment("计划结束时间").
			Optional(),
		field.Time("actual_start_date").
			Comment("实际开始时间").
			Optional(),
		field.Time("actual_end_date").
			Comment("实际结束时间").
			Optional(),
		field.Text("implementation_plan").
			Comment("实施计划").
			Optional(),
		field.Text("rollback_plan").
			Comment("回滚计划").
			Optional(),
		field.JSON("affected_cis", []string{}).
			Comment("受影响的配置项").
			Optional(),
		field.JSON("related_tickets", []string{}).
			Comment("相关工单").
			Optional(),
		field.Time("created_at").
			Comment("创建时间").
			Default(time.Now),
		field.Time("updated_at").
			Comment("更新时间").
			Default(time.Now).
			UpdateDefault(time.Now),
	}
}

// Edges of the Change.
func (Change) Edges() []ent.Edge {
	return nil
}

File: itsm-frontend/src/types/biz/change.ts
/**
 * 变更管理类型定义
 */

import type {
  ChangeStatus,
  ChangeType,
  ChangePriority,
  ChangeImpact,
  ChangeRisk,
} from '@/constants/change';

// 变更实体
export interface Change {
  id: number;
  title: string;
  description: string;
  justification: string;
  type: ChangeType;
  status: ChangeStatus;
  priority: ChangePriority;
  impactScope: ChangeImpact;
  riskLevel: ChangeRisk;
  assigneeId?: number;
  assigneeName?: string;
  createdBy: number;
  createdByName: string;
  tenantId: number;
  plannedStartDate?: string;
  plannedEndDate?: string;
  actualStartDate?: string;
  actualEndDate?: string;
  implementationPlan: string;
  rollbackPlan: string;
  affectedCis?: string[];
  relatedTickets?: string[];
  createdAt: string;
  updatedAt: string;
}

// 审批记录
export interface ApprovalRecord {
  id: number;
  changeId: number;
  approverId: number;
  approverName: string;
  status: ChangeStatus;
  comment?: string;
  approvedAt?: string;
  createdAt: string;
}

// 审批链/工作流项
export interface ApprovalChainItem {
  id: number;
  level: number;
  approverId: number;
  approverName: string;
  role: string;
  status: string;
  isRequired: boolean;
}

// 风险评估
export interface RiskAssessment {
  id: number;
  changeId: number;
  riskLevel: ChangeRisk;
  riskDescription: string;
  impactAnalysis: string;
  mitigationMeasures: string;
  contingencyPlan?: string;
  riskOwner: string;
  riskReviewDate?: string;
}

// 创建变更请求
export interface CreateChangeRequest {
  title: string;
  description: string;
  justification: string;
  type: ChangeType;
  priority: ChangePriority;
  impactScope: ChangeImpact;
  riskLevel: ChangeRisk;
  plannedStartDate?: string;
  plannedEndDate?: string;
  implementationPlan: string;
  rollbackPlan: string;
  affectedCis?: string[];
  relatedTickets?: string[];
}

// 列表查询参数
export interface ChangeQuery {
  page?: number;
  pageSize?: number;
  status?: string;
  type?: string;
  search?: string;
}

// 列表响应
export interface ChangeListResponse {
  changes: Change[];
  total: number;
  page: number;
  size: number;
}

// 统计响应
export interface ChangeStats {
  total: number;
  pending: number;
  approved: number;
  inProgress: number;
  completed: number;
  rolledBack: number;
  rejected: number;
  cancelled: number;
}
